src/Generation/ResourcesGenerator.php

```diff
@@ -524,45 +524,33 @@ private static function getQueryParams(MethodDetails $method): Vector
      *
      * @return array<string, string> Fields to autopopulate in GAX if unset by user
      */
-    private static function getUuid4FieldsToAutoPopulate(
+    private static function getAutoPopulatedUuid4Fields(
         MethodDetails $method,
-        ServiceDetails $serviceDetails
+        ServiceDetails $service
     ): array {
-        $autoPopulatedFields = Vector::new([]);
-
-        $methodSettings = $serviceDetails->serviceYamlConfig->methodSettings;
-        foreach ($methodSettings as $methodSetting) {
-            $autoPopulatedFieldNames = iterator_to_array($methodSetting->getAutoPopulatedFields());
-            if (!empty($autoPopulatedFieldNames)) {
-
-                // Check if the $methodSetting is for this particular $method
-                $splitMethodName = explode('.', $methodSetting->getSelector());
-                $methodName = $splitMethodName[array_key_last($splitMethodName)];
-                if ($methodName !== $method->name) {
-                    continue;
-                }
+        $result = Map::new();
 
-                // Process only if the method is unary
-                if (!$method->isStreaming()) {
-                    $autoPopulatedFields = $method->optionalFields
-                        ->filter(
-                            // Filter out field names to autopopulate from all fields of the method
-                            fn ($f) => in_array($f->name, $autoPopulatedFieldNames)
-                        )
-                        ->filter(
-                            // Filter for optional field and uuid4 format
-                            fn ($f) => !$f->isRequired && $f->isUuid4
-                        );
-                }
-            }
+        if ($method->isStreaming()) {
+            return [];
+        }
+
+        $methodSetting = $service->serviceYamlConfig->methodSettings
+            ->filter(function ($x) use ($method) {
+                $splitName = explode('.', $x->getSelector());
+                return $method->name == $splitName[array_key_last($splitName)];
+            })->firstOrNull();
+
+        if (!is_null($methodSetting) && count($methodSetting->getAutoPopulatedFields())) {
+            $fieldNames = iterator_to_array($methodSetting->getAutoPopulatedFields());
+            $result = $method->optionalFields
+                ->filter(fn ($x) => in_array($x->name, $fieldNames))
+                ->filter(fn ($x) => $x->isUuid4)
+                ->toMap(fn ($x) => $x->camelName, fn ($x) => 'UUID4');
         }
-        $autoPopulatedFields = $autoPopulatedFields->map(fn ($x) => $x->camelName)->toArray();
 
-        // Convert array of field_names to associative array of
-        // field_name => field_type
         return array_combine(
-            $autoPopulatedFields,
-            array_fill(0, count($autoPopulatedFields), 'UUID4')
+            $result->keys()->toArray(),
+            $result->values()->toArray()
```

Comment:
This feels like something we should have in the `Map` API e.g. `toArray()` which just returns the underlying `$data` property (maybe a copy of it) which is itself an associative array...WDYT? Would you mind adding/using that?

Reply:
Yeah I totally support this feeling. I was skeptical on calling this out because the Vector, Map, Set...these API feels like they mimic C++ style data structures to some extent and hence their design is kind of same for all languages (I may be wrong here as I never verified this fact).

Anyways, now that you also called out, I've extracted the a `toArray()` code to `Map` api.